src/libdredd/src/mutation_replace_expr.cc

```diff
@@ -0,0 +1,221 @@
+// Copyright 2022 The Dredd Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "libdredd/mutation_replace_expr.h"
+
+#include <cassert>
+#include <sstream>
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/Type.h"
+#include "clang/Basic/LangOptions.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+#include "libdredd/util.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace dredd {
+dredd::MutationReplaceExpr::MutationReplaceExpr(const clang::Expr& expr)
+    : expr_(expr) {}
+
+std::string MutationReplaceExpr::GetFunctionName() const {
+  std::string result = "__dredd_replace_expr_";
+
+  // A string corresponding to the expression forms part of the name of the
+  // mutation function, to differentiate mutation functions for different
+  // types
+  if (expr_.getType()->isBooleanType()) {
+    result += "Bool";
+  } else if (expr_.getType()->isIntegerType()) {
+    result += "Int";
+  } else if (expr_.getType()->isFloatingType()) {
+    result += "Float";
+  } else {
+    assert(false && "Unsupported opcode");
+  }
+
+  if (expr_.isLValue()) {
+    result += "_lvalue";
+  }
```

Comment:
Please make sure that ultimately there is a single file test case that relies on this distinction - i.e. one for which there would be a name clash if `lvalue` was not present.

Reply:
Added a single file test case.